### PR TITLE
Correctly compute gradients for torch backend

### PR DESCRIPTION
**Pull request by zimmerrol**
Currently, the computation of gradients using `value_and_grad` for torch tensors has the problem of storing the gradients in the model. This can become an issue for adversarial training. This PR is meant to solve this issue.

This PR also updates the required versions of tensorflow and black to solve some incompatibility issues with packages they depend on.

### CODE REVIEW

**Comment by zimmerrol** [comment]
@jonasrauber Can you please review this PR?
